### PR TITLE
fix crash when the expression exceed the depth

### LINKED ISSUE #3393: Limit the depth of the expression, report an error when it exceeds but not crash

### PR DESCRIPTION
**Pull request by heroicNeZha**
#### What type of PR is this?
- [x] bug
- [ ] feature
- [ ] enhancement

#### What does this PR do?
fix crash When the expression's depth exceeds the maximum limit
（The depth is related to the stack size. This pr limited 815 is calculated based on the stack size of the development machine）

#### Which issue(s)/PR(s) this PR relates to?
fixed #3393 
  
#### Checklist：
- [ ] Documentation affected （Please add the label if documentation needs to be modified.)
- [ ] Incompatible （If it is incompatible, please describe it and add corresponding label.）
- [ ] Need to cherry-pick （If need to cherry-pick to some branches, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

#### Release notes：
Please confirm whether to reflect in release notes and how to describe:
>                                                                 `

### CODE REVIEW

File: src/graph/visitor/DeduceTypeVisitor.cpp [hunk 1]
```diff
@@ -94,6 +94,16 @@ static const std::unordered_map<Value::Type, Value> kConstantValues = {
   }                                                                  \
   type_ = detectVal.type()
 
+#define CHECK_DEPTH(depth)                                 \
```
Comment (Shylock-Hg) [inline src/graph/visitor/DeduceTypeVisitor.cpp:97]: Better to replace by method.

**Comment by Shylock-Hg** [comment]
And maybe not all expression will be checked by `DeduceTypeVisitor`.

File: src/graph/visitor/DeduceTypeVisitor.cpp (hunk 1 above)
Comment (heroicNeZha) [reply src/graph/visitor/DeduceTypeVisitor.cpp:97]: Sorry I am new in c++,  I wanna return in the macro and don't know how to replace it with a method

Comment (Shylock-Hg) [reply src/graph/visitor/DeduceTypeVisitor.cpp:97]: As bellow:
```c++
checkDepth();
if (status_.ok() == false) {
    return;
}
```

Comment (heroicNeZha) [reply src/graph/visitor/DeduceTypeVisitor.cpp:97]: thanks！

**Comment by heroicNeZha** [comment]
> And maybe not all expression will be checked by `DeduceTypeVisitor`.

The `DeduceTypeVisitor` extends pure virtual class `ExprVisitor` which seems contains all expressions

And I didn't check non-recursive expressions in `DeduceTypeVisitor`

File: src/graph/visitor/DeduceTypeVisitor.cpp [hunk 2]
```diff
@@ -116,6 +116,10 @@ void DeduceTypeVisitor::visit(ConstantExpression *expr) {
 }
 
 void DeduceTypeVisitor::visit(UnaryExpression *expr) {
+  checkDepth();
+  if (!status_.ok()) {
+    return;
```
Comment (yixinglu) [inline src/graph/visitor/DeduceTypeVisitor.cpp:121]: should always recover depth when exit current function:
```c++
checkDepth();
SCOPED_EXIT {
  recoverDepth();
}
```

However I also don't think we should do this thing in this visitor.

Comment (heroicNeZha) [reply src/graph/visitor/DeduceTypeVisitor.cpp:121]: ok I will create a new visitor

**Comment by critical27** [comment]
Generally LGTM

File: src/graph/visitor/CheckDepthVisitor.cpp [hunk 3]
```diff
@@ -0,0 +1,251 @@
+/* Copyright (c) 2020 vesoft inc. All rights reserved.
```
Comment (Aiee) [inline src/graph/visitor/CheckDepthVisitor.cpp:1]: 2021

File: src/graph/visitor/CheckDepthVisitor.h [hunk 4]
```diff
@@ -0,0 +1,99 @@
+/* Copyright (c) 2020 vesoft inc. All rights reserved.
```
Comment (Aiee) [inline src/graph/visitor/CheckDepthVisitor.h:1]: ditto

File: tests/tck/features/expression/Depth.feature [hunk 5]
```diff
@@ -0,0 +1,160 @@
+# Copyright (c) 2021 vesoft inc. All rights reserved.
+#
+# This source code is licensed under Apache 2.0 License.
+@li
```
Comment (Aiee) [inline tests/tck/features/expression/Depth.feature:4]: delete testing mark

File: src/graph/visitor/CheckDepthVisitor.h [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ * This source code is licensed under Apache 2.0 License.
+ */
+
+#ifndef GRAPH_VISITOR_CHECKDEPTHVISITOR_H_
+#define GRAPH_VISITOR_CHECKDEPTHVISITOR_H_
```
Comment (Aiee) [inline src/graph/visitor/CheckDepthVisitor.h:7]: use `#pragma once`

File: src/graph/validator/FindPathValidator.cpp [hunk 7]
```diff
@@ -46,6 +46,7 @@ Status FindPathValidator::validateWhere(WhereClause* where) {
   }
   where->setFilter(ExpressionUtils::rewriteLabelAttr2EdgeProp(expr));
   auto filter = where->filter();
+  NG_RETURN_IF_ERROR(checkExprDepth(filter));
 
   auto typeStatus = deduceExprType(filter);
```
Comment (yixinglu) [inline src/graph/validator/FindPathValidator.cpp:51]: It seems that you only need to do this `checkExprDepth()` inside `deduceExprType()`.  So you don't have to add this check everywhere.

Comment (heroicNeZha) [reply src/graph/validator/FindPathValidator.cpp]: done

File: src/graph/visitor/CheckDepthVisitor.cpp [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+  const_cast<LabelExpression *>(expr->left())->accept(this);
+  if (!ok()) return;
+  const_cast<ConstantExpression *>(expr->right())->accept(this);
+}
+
+void CheckDepthVisitor::visit(FunctionCallExpression *expr) {
+  checkDepth();
+  SCOPE_EXIT { recoverDepth(); };
+  if (!ok()) return;
+  for (auto &arg : expr->args()->args()) {
```
Comment (jiayuehua) [inline src/graph/visitor/CheckDepthVisitor.cpp:108]: consider use any_of

File: src/graph/visitor/CheckDepthVisitor.cpp [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+  if (!ok()) return;
+  expr->left()->accept(this);
+  if (!ok()) return;
+  expr->right()->accept(this);
+}
+
+void CheckDepthVisitor::visit(SubscriptExpression *expr) {
+  checkDepth();
+  SCOPE_EXIT { recoverDepth(); };
+  if (!ok()) return;
```
Comment (jiayuehua) [inline src/graph/visitor/CheckDepthVisitor.cpp:69]: consider refactor to two function that return bool and use: fa() and fb();

Comment (heroicNeZha) [reply src/graph/visitor/CheckDepthVisitor.cpp:69]: this pr is closed

**Pull request closed without merging**